server/http.go

```diff
@@ -29,11 +32,186 @@ func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Req
 		p.handleWebhook(w, r)
 	case "/api/v1/meetings":
 		p.handleStartMeeting(w, r)
+	case "/oauth2/connect":
+		p.connectUserToZoom(w, r)
+	case "/external/oauth2/connect":
+		p.connectExternalUserToZoom(w, r)
+	case "/oauth2/complete":
+		p.completeUserOAuthToZoom(w, r)
 	default:
 		http.NotFound(w, r)
 	}
 }
 
+func (p *Plugin) connectUserToZoom(w http.ResponseWriter, r *http.Request) {
+	userID := r.Header.Get("Mattermost-User-ID")
+	if userID == "" {
+		http.Error(w, "Not authorized", http.StatusUnauthorized)
+		return
+	}
+
+	channelID := r.URL.Query().Get("channelID")
+
+	if channelID == "" {
+		http.Error(w, "Not authorized", http.StatusUnauthorized)
+		return
+	}
+
+	conf, err := p.getOAuthConfig()
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+
+	key := fmt.Sprintf("%v_%v", model.NewId()[0:15], userID)
+	state := fmt.Sprintf("%v_%v", key, channelID)
+
+	appErr := p.API.KVSet(key, []byte(state))
+	if appErr != nil {
+		http.Error(w, appErr.Error(), http.StatusInternalServerError)
+	}
+
+	url := conf.AuthCodeURL(state, oauth2.AccessTypeOffline)
+	http.Redirect(w, r, url, http.StatusFound)
+}
+
+func (p *Plugin) connectExternalUserToZoom(w http.ResponseWriter, r *http.Request) {
+	channelID := r.URL.Query().Get("channelID")
+
+	if channelID == "" {
+		http.Error(w, "Missing channel ID", http.StatusBadRequest)
+		return
+	}
+
+	userID := r.URL.Query().Get("userID")
```

Comment:
Doesn't this allow an unauthenticated user to connect their Zoom account to *any* Mattermost account, including those that belong to other people? Instead of passing the user ID directly, I think we should be generating secret tokens that are mapped back to user IDs on the server.

Reply:
I can do that. Here the idea:
I create the Secret Token within at the same time as the post, and add it to the URL (`something.com/plugin/external/oauth?channelID=xxx&userID=xxx&secretToken=xxx`). The Secret Token will create a short lived (5 minutes?) key in the KV store that points to the user ID. The secret token will be included in the OAuth state, and checked on the way back of the connection.

Does this make sense?

Reply:
Yeah, that sounds good. Though wouldn't that also allow getting rid of both `channelID` and `userID` in the URL if we just store them in the KV store instead? So you could just pass `something.com/plugin/external/oauth?secretToken=xxx` without the other two parameters.